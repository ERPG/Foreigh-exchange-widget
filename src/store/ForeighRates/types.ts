import { ForeighRates } from '../../models/ForeighRates';

export const GET_FOREIGH_RATES = 'GET_FOREIGH_RATES';
export const GENERETATE_FOREIGH_RATES_ERROR = 'GENERETATE_FOREIGH_RATES_ERROR';

export interface GetForeighRatesAction {
  type: typeof GET_FOREIGH_RATES;
  payload: ForeighRates;
}

export interface GenerateForeighRatesErrorAction {
  type: typeof GENERETATE_FOREIGH_RATES_ERROR;
  error: Error;
}

export type ForeighRatesActionTypes =
  | GetForeighRatesAction
  | GenerateForeighRatesErrorAction;
