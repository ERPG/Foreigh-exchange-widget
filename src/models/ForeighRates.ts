export enum Currencies {
  EUR = 'EUR',
  USD = 'USD',
  GBP = 'GBP',
}

export type Rates<R> = { [key in keyof typeof Currencies]: R };

export type CurrencyRate<T> = { currency: Currencies; rate: T };

export interface IForeighRatesPayload {
  base: Currencies;
  rates: Rates<string>;
}

export class ForeighRates {
  private readonly _base: Currencies;
  private readonly _rates: Rates<string>;

  constructor(foreighRates: IForeighRatesPayload) {
    this._rates = foreighRates.rates;
    this._base = foreighRates.base;
  }

  get base(): Currencies {
    return this._base;
  }

  get rates(): Rates<string> {
    return this._rates;
  }

  public getExchangeRatesArr() {
    let newRatesArr: CurrencyRate<number>[] = [];
    for (const key in this._rates) {
      newRatesArr.push({
        currency: key as Currencies,
        // == CAMBIAR EL ANY ==
        rate: (this._rates as any)[key],
      });
    }
    return newRatesArr;
  }
}
