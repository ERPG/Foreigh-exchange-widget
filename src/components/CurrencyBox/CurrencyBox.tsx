import { Paper, SelectChangeEvent } from '@mui/material';
import Input from '../../common/components/Input/Input';
import Select from '../../common/components/Select/Select';
import { CurrencyRate } from '../../models/ForeighRates';

export interface ICurrencyBoxProps {
  id: string;
  currencyRates: CurrencyRate<number>[] | undefined;
  amount: number;
  handleAmount: (e: React.ChangeEvent<HTMLInputElement>) => void;
  selectValue: number;
  handleSelectChange: (event: SelectChangeEvent<string>) => void;
  inputLabel: string;
}

const CurrencyBox = ({
  id,
  inputLabel,
  currencyRates,
  amount,
  handleAmount,
  selectValue,
  handleSelectChange,
}: ICurrencyBoxProps) => {
  return (
    <Paper
      style={{
        textAlign: 'center',
        height: 80,
        lineHeight: '60px',
        margin: 16,
        display: 'flex',
        justifyContent: 'space-around',
        alignItems: 'center',
      }}
    >
      <div>
        <Input
          label={inputLabel}
          name={`${id}-input`}
          value={amount}
          onChange={handleAmount}
        />
      </div>
      <div style={{ marginTop: 16 }}>
        <Select
          name={`${id}-select`}
          items={currencyRates}
          value={selectValue}
          onChange={handleSelectChange}
        />
      </div>
    </Paper>
  );
};

export default CurrencyBox;
