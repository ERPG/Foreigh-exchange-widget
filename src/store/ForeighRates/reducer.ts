import { AppState } from '..';
import { ForeighRates } from '../../models/ForeighRates';
import {
  ForeighRatesActionTypes,
  GENERETATE_FOREIGH_RATES_ERROR,
  GET_FOREIGH_RATES,
} from './types';

export interface IForeighRatesState {
  item?: ForeighRates;
  error?: Error;
}

export const INITIAL_STATE: IForeighRatesState = { item: undefined };

export function ForeighRatesReducer(
  state = INITIAL_STATE,
  action: ForeighRatesActionTypes
): IForeighRatesState {
  switch (action.type) {
    case GET_FOREIGH_RATES:
      return { ...state, item: action.payload };
    case GENERETATE_FOREIGH_RATES_ERROR:
      return { ...state, error: action.error };
    default:
      return state;
  }
}

export const selectForeighRates = (state: AppState): ForeighRates | undefined =>
  state.item;
