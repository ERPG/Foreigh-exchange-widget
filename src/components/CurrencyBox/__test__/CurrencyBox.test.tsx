import { fireEvent } from '@testing-library/react';
import { generateAppStateMock } from '../../../../__tests__/__fixtures__/generateAppStateMock';
import CurrencyBox from '../CurrencyBox';
import { Currencies } from '../../../models/ForeighRates';

import { renderStoreHelper } from '../../../../__tests__/__utils__/renderStoreHelper';

describe('CurrencyBox component', () => {
  const mockCallback = jest.fn();
  const defaultProps = {
    id: 'mock-id',
    currencyRates: [
      { currency: Currencies.EUR, rate: 1 },
      { currency: Currencies.USD, rate: 1 },
    ],
    amount: 10,
    handleAmount: mockCallback,
    selectValue: 0,
    handleSelectChange: mockCallback,
    inputLabel: 'mock-label',
  };

  it('Should render component without errors', () => {
    const { container } = renderStoreHelper({
      component: <CurrencyBox {...defaultProps} />,
      rootState: generateAppStateMock(),
    });

    expect(container).toMatchSnapshot();
  });

  it('Should call amount on input change', () => {
    const { getByDataFx } = renderStoreHelper({
      component: <CurrencyBox {...defaultProps} />,
      rootState: generateAppStateMock(),
    });

    const inputEl = getByDataFx('fx-amount-input');
    expect(inputEl).toBeDefined();

    fireEvent.change(inputEl, { target: { value: '123' } });
    expect(mockCallback).toBeCalled();
  });

  it('Should call currency selection on input change', () => {
    const { getByDataFx } = renderStoreHelper({
      component: <CurrencyBox {...defaultProps} />,
      rootState: generateAppStateMock(),
    });

    const selectEl = getByDataFx('fx-currency-selection');
    expect(selectEl).toBeDefined();

    fireEvent.change(selectEl, { target: { value: '1' } });
    expect(mockCallback).toBeCalled();
  });
});
