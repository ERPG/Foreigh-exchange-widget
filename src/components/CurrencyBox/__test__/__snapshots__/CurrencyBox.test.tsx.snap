// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyBox component Should render component without errors 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1ps6pg7-MuiPaper-root"
    style="text-align: center; height: 80px; line-height: 60px; margin: 16px; display: flex; justify-content: space-around; align-items: center;"
  >
    <div>
      <div
        class="MuiFormControl-root MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-marginDense MuiFormLabel-filled"
          data-shrink="true"
          for="currency-amount"
          id="currency-amount-label"
        >
          mock-label
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense MuiInput-marginDense"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
            data-fx="fx-amount-input"
            id="currency-amount"
            inputmode="numeric"
            name="mock-id-input"
            type="text"
            value="10"
          />
        </div>
      </div>
    </div>
    <div
      style="margin-top: 16px;"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth css-1v3hyuc-MuiFormControl-root"
      >
        <div
          class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl makeStyles-root-1 MuiSelect-root css-1480iag-MuiInputBase-root-MuiInput-root"
        >
          <div
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-label="Without label"
            aria-labelledby="mui-component-select-mock-id-select"
            class="MuiSelect-select MuiSelect-standard MuiInput-input MuiInputBase-input css-1rxz5jq-MuiSelect-select-MuiInputBase-input-MuiInput-input"
            id="mui-component-select-mock-id-select"
            role="button"
            tabindex="0"
          >
            <span
              class="notranslate"
            >
              â€‹
            </span>
          </div>
          <input
            aria-hidden="true"
            class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
            data-fx="fx-currency-selection"
            name="mock-id-select"
            tabindex="-1"
            value="0"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconStandard css-pqjvzy-MuiSvgIcon-root-MuiSelect-icon"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
